service/dap/server.go

```diff
@@ -1254,6 +1258,81 @@ func (s *Server) onSetFunctionBreakpointsRequest(request *dap.SetFunctionBreakpo
 	s.send(response)
 }
 
+const instructionBpPrefix = "instructionBreakpoint"
+
+func (s *Server) onSetInstructionBreakpointsRequest(request *dap.SetInstructionBreakpointsRequest) {
+	// Get all existing function breakpoints.
+	existingBps := s.getMatchingBreakpoints(instructionBpPrefix)
+	bpAdded := make(map[string]struct{}, len(existingBps))
+	for _, bp := range existingBps {
+		existingBps[bp.Name] = bp
+	}
+
+	breakpoints := make([]dap.Breakpoint, len(request.Arguments.Breakpoints))
```

Comment:
It seems to me that this loop as well as the next exists almost identically also in onSetFunctionBreakpointsRequest and onSetBreakpointsRequest. Wouldn't it be better to refactor it to a common function, like:

```
breakpoints := updateExistingBreakpoints(len(request.Arguments.Brekapoints), func(i int) (name, cond, hitCond string) {
	want := &request.Arguments.Breakpoints[i]
	return fmt.Sprintf("%s PC=%s", instructionBpPrefix, want.InstructionReference), want.Cond, want.HitCond
})
```

where breakpoints is `[]*dap.Breakpoint` and is nil for breakpoints that the third, final loop needs to add (because they are new).

There is one difference that I noticed that this and SetFunctionBreakpointsRequest calls toClientPath but SetBreakpointsRequest doesn't.

Reply:
There is a lot of shared code here. Thanks for the suggestion on how to refactor this. I did a slightly modified refactor to allow the three functions to share all of the logic (besides location computation).